File: vb/VS2013/netframework/Modules/10.API/22.Virtual Mode/CellReader.vb
Imports System.Text
Imports FlexCel.Core

Namespace VirtualMode
	'A simple cell reader that will get the values from FlexCel and put them into a grid.
	Friend Class CellReader
		Private Only50Rows As Boolean
		Private CellData As SparseCellArray
		Private FormatValues As Boolean
		Private SheetToRead As Integer
		Public StartSheetSelect As Date
		Public EndSheetSelect As Date

		Public Sub New(ByVal aOnly50Rows As Boolean, ByVal aCellData As SparseCellArray, ByVal aFormatValues As Boolean)
			Only50Rows = aOnly50Rows
			CellData = aCellData
			FormatValues = aFormatValues
		End Sub

		Public Sub OnStartReading(ByVal sender As Object, ByVal e As VirtualCellStartReadingEventArgs)
			StartSheetSelect = Date.Now
			Using SheetSelector As New SheetSelectorForm(e.SheetNames)

				If Not SheetSelector.Execute() Then
					EndSheetSelect = Date.Now
					e.NextSheet = Nothing 'stop reading
					Return
				End If

				EndSheetSelect = Date.Now
				e.NextSheet = SheetSelector.SelectedSheet
				SheetToRead = SheetSelector.SelectedSheetIndex + 1
			End Using
		End Sub

		Public Sub OnCellRead(ByVal sender As Object, ByVal e As VirtualCellReadEventArgs)
			If Only50Rows AndAlso e.Cell.Row > 50 Then
				e.NextSheet = Nothing 'Stop reading all sheets.
				Return
			End If

			If e.Cell.Sheet <> SheetToRead Then
				e.NextSheet = Nothing 'Stop reading all sheets.
				Return
			End If

			If FormatValues Then
				Dim Clr As TUIColor = Color.Empty
				CellData.AddValue(e.Cell.Row, e.Cell.Col, TFlxNumberFormat.FormatValue(e.Cell.Value, CType(sender, ExcelFile).GetFormat(e.Cell.XF).Format, Clr, (CType(sender, ExcelFile))))
			Else
				CellData.AddValue(e.Cell.Row, e.Cell.Col, Convert.ToString(e.Cell.Value))
			End If
		End Sub
	End Class
End Namespace

File: csharp/VS2015/netframework/Modules/10.API/22.Virtual Mode/CellReader.cs
using System;
using System.Collections.Generic;
using System.Text;
using FlexCel.Core;
using System.Drawing;

namespace VirtualMode
{
    //A simple cell reader that will get the values from FlexCel and put them into a grid.
    class CellReader
    {
        private bool Only50Rows;
        private SparseCellArray CellData;
        private bool FormatValues;
        private int SheetToRead;
        public DateTime StartSheetSelect;
        public DateTime EndSheetSelect;

        public CellReader(bool aOnly50Rows, SparseCellArray aCellData, bool aFormatValues)
        {
            Only50Rows = aOnly50Rows;
            CellData = aCellData;
            FormatValues = aFormatValues;
        }

        public void OnStartReading(object sender, VirtualCellStartReadingEventArgs e)
        {
            StartSheetSelect = DateTime.Now;
            using (SheetSelectorForm SheetSelector = new SheetSelectorForm(e.SheetNames))
            {

                if (!SheetSelector.Execute())
                {
                    EndSheetSelect = DateTime.Now;
                    e.NextSheet = null; //stop reading
                    return;
                }

                EndSheetSelect = DateTime.Now;
                e.NextSheet = SheetSelector.SelectedSheet;
                SheetToRead = SheetSelector.SelectedSheetIndex + 1;
            }
        }

        public void OnCellRead(object sender, VirtualCellReadEventArgs e)
        {
            if (Only50Rows && e.Cell.Row > 50)
            {
                e.NextSheet = null; //Stop reading all sheets.
                return;
            }

            if (e.Cell.Sheet != SheetToRead)
            {
                e.NextSheet = null; //Stop reading all sheets.
                return;
            }

            if (FormatValues)
            {
                TUIColor Clr = Color.Empty;
                CellData.AddValue(e.Cell.Row, e.Cell.Col,
                   TFlxNumberFormat.FormatValue(e.Cell.Value,
                   ((ExcelFile)sender).GetFormat(e.Cell.XF).Format, ref Clr, ((ExcelFile)sender)));
            }
            else
            {
                CellData.AddValue(e.Cell.Row, e.Cell.Col, Convert.ToString(e.Cell.Value));
            }
        }
    }
}
